observe/observe.go

```diff
@@ -0,0 +1,110 @@
+// Package observe provides functions
+// that help with setting tracing/metrics
+// in cloud providers, mainly GCP.
+package observe
+
+import (
+	"context"
+	"os"
+
+	traceapi "cloud.google.com/go/trace/apiv2"
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	"contrib.go.opencensus.io/exporter/stackdriver/monitoredresource"
+	"go.opencensus.io/exporter/prometheus"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+)
+
+// NewStackDriverExporter will return the tracing and metrics through
+// the stack driver exporter, if exists in the underlying platform.
+// If exporter is registered, it returns the exporter so you can register
+// it and ensure to call Flush on termination.
+func NewStackDriverExporter(projectID string, onErr func(error)) (*stackdriver.Exporter, error) {
+	_, svcName, svcVersion := GetServiceInfo()
+	opts := getSDOpts(projectID, svcName, svcVersion, onErr)
+	if opts == nil {
+		return nil, nil
+	}
+	return stackdriver.NewExporter(*opts)
+}
+
+// NewPrometheusExporter return a prometheus Exporter for OpenCensus.
+func NewPrometheusExporter(opts prometheus.Options) (*prometheus.Exporter, error) {
+	return prometheus.NewExporter(opts)
+}
+
+// GoogleProjectID returns the GCP Project ID
+// that can be used to instantiate various
+// GCP clients such as Stack Driver.
+func GoogleProjectID() string {
+	return os.Getenv("GOOGLE_CLOUD_PROJECT")
+}
+
+// IsGAE tells you whether your program is running
+// within the App Engine platform.
+func IsGAE() bool {
+	return os.Getenv("GAE_DEPLOYMENT_ID") != ""
+}
+
+// GetGAEInfo returns the GCP Project ID,
+// the service, and the version of the application.
+func GetGAEInfo() (projectID, service, version string) {
+	return GoogleProjectID(),
+		os.Getenv("GAE_SERVICE"),
+		os.Getenv("GAE_VERSION")
+}
+
+// GetServiceInfo returns the GCP Project ID,
+// the service name and version (GAE or through
+// GAE_SERVICE/GAE_VERSION env vars). Note
+// that GAE_SERVICE/GAE_VERSION are not standard but
```

Comment:
The `GAE_SERVICE/VERSION` vars actually are standard. Those get injected into the env by the platform.

I was talking about the `SERVICE_NAME/SERVICE_VERSION` env vars as not being the standard. We should make sure we mention those as well.

Reply:
For the GAE vars, I wonder if we should provide a link to the docs? https://cloud.google.com/appengine/docs/standard/go111/runtime#environment_variables